src/dmd/statementsem.d

```diff
@@ -3161,7 +3161,10 @@ else
                 rs.exp = inferType(rs.exp, fld.treq.nextOf().nextOf());
 
             rs.exp = rs.exp.expressionSemantic(sc);
-            rs.exp.checkSharedAccess(sc);
+            if(inferRef || !tf.isref)
```

Comment:
Why if `inferRef`? What case does that represent?

Reply:
inferRef means `auto ref`.
In which case we don't want to allow unsolicited shared access.

Reply:
Why should auto ref not infer ref?

Reply:
> Because then code you have no control over affects the behavior of shared

How so, it should just do the appropriate `shared` checks after inferring ref/value. Why should there any differences between `foo` and `bar`?

```
void main()
{
    shared bool b = void;
    foo(b);
    bar(b);
}

void foo(ref shared bool b) { atomicStore(b, false); }
void bar()(auto ref shared bool b) { atomicStore(b, false); }
```

Reply:
> Because the ref was explicitly written and will always be ref.

That will be the case for `bar` as well because reading actual shared data would't compile at alll.

> shared is all about making you AWARE of problem with sharing data

When using `auto ref` you already have to make the defensive assumption that it infers `ref`, not the other way round. Raising awareness is certainly a good intention but it doesn't justify odd special cases IMHO.

Anyway, this can also be deferred into another PR as the current change provides a net benefit already.

Reply:
I agree that the exception is weird. If `auto ref` infers `ref`, then the call should work normally.